src/artifact.rs

```diff
@@ -447,10 +447,34 @@ impl Artifact {
 
     }
 
+    /// Get set of non-import declarations that have not been defined. This must be an empty set in
+    /// order to `emit` the artifact.
+    pub fn undefined_symbols(&self) -> Vec<String> {
```

Comment:
At first I was worried that this was going to allocate a `Vec` of allocated `String` all the time, but it turns out that when everything is working as intended, the `Vec` is empty, so nothing is allocated :-).